### ISSUE TITLE
Precedence defintion

### ISSUE THREAD

**Issue by thekid** [xp-runners/reference#11]
The precedence of class and module paths is currently undefined. 
## This is the suggested class path precedence:
1. **Overlays**, if any
2. **XP Core**, which is loaded from whichever of the following comes first:
   - A file entry from the local .pth files, e.g. `./vendor/autoload.php`, which includes `__xp.php`
   - A directory or xar archive containing `__xp.php`
   - The global `use` directive, either from `USE_XP` environment variable or xp.ini
3. **Local module** (basically, all entries in the local .pth files)
4. **Any other class path entries** from `-cp`
5. **Modules from `-m`**
6. **The current directory**

**Comment by thekid** [comment]
Example

```sh
Timm@slate ~/devel/xp/command [feature/create]
$ cat class.pth
!src/main/php/
src/test/php/
src/test/resources/

Timm@slate ~/devel/xp/command [feature/create]
$ xp -m ../sequence/ -cp ../inject/src/main/php/ -v
SCAN .
`- CORE? .\src\main\php\lang.base.php -> no
`- PATH class.pth
   `- ADD overlay C:\tools\cygwin\home\Timm\devel\xp\command\src\main\php\
   `- ADD local C:\tools\cygwin\home\Timm\devel\xp\command\src\test\php\
   `- ADD local C:\tools\cygwin\home\Timm\devel\xp\command\src\test\resources\

SCAN C:\tools\cygwin\home\Timm\devel\xp\core
`- CORE? C:\tools\cygwin\home\Timm\devel\xp\core\src\main\php\lang.base.php -> yes
`- PATH boot.pth
   `- ADD core C:\tools\cygwin\home\Timm\devel\xp\core\src\main\php\
   `- ADD core C:\tools\cygwin\home\Timm\devel\xp\core\src\test\php\
   `- ADD core C:\tools\cygwin\home\Timm\devel\xp\core\src\main\resources\
   `- ADD core C:\tools\cygwin\home\Timm\devel\xp\core\src\test\resources\

SCAN ../sequence/
`- PATH class.pth
   `- ADD local C:\tools\cygwin\home\Timm\devel\xp\sequence\src\main\php\
   `- ADD local C:\tools\cygwin\home\Timm\devel\xp\sequence\src\test\php\
`- PATH glue.pth
   `- ADD local C:\tools\cygwin\home\Timm\devel\xp\collections\src\main\php\

ADD ../inject/src/main/php/

LOCAL .

array(4) {
  ["base"]=>
  string(66) "C:\tools\cygwin\home\Timm\devel\xp\core\src\main\php\lang.base.php"
  ["overlay"]=>
  array(1) {
    [0]=>
    string(56) "C:\tools\cygwin\home\Timm\devel\xp\command\src\main\php\"
  }
  ["core"]=>
  array(4) {
    [0]=>
    string(53) "C:\tools\cygwin\home\Timm\devel\xp\core\src\main\php\"
    [1]=>
    string(53) "C:\tools\cygwin\home\Timm\devel\xp\core\src\test\php\"
    [2]=>
    string(59) "C:\tools\cygwin\home\Timm\devel\xp\core\src\main\resources\"
    [3]=>
    string(59) "C:\tools\cygwin\home\Timm\devel\xp\core\src\test\resources\"
  }
  ["local"]=>
  array(7) {
    [0]=>
    string(56) "C:\tools\cygwin\home\Timm\devel\xp\command\src\test\php\"
    [1]=>
    string(62) "C:\tools\cygwin\home\Timm\devel\xp\command\src\test\resources\"
    [2]=>
    string(57) "C:\tools\cygwin\home\Timm\devel\xp\sequence\src\main\php\"
    [3]=>
    string(57) "C:\tools\cygwin\home\Timm\devel\xp\sequence\src\test\php\"
    [4]=>
    string(60) "C:\tools\cygwin\home\Timm\devel\xp\collections\src\main\php\"
    [5]=>
    string(55) "C:\tools\cygwin\home\Timm\devel\xp\inject\src\main\php\"
    [6]=>
    string(43) "C:\tools\cygwin\home\Timm\devel\xp\command\"
  }
}
```

**Comment by thekid** [comment]
## Example for a local installation

```
Timm@slate ~/devel/xp/handlebars [master]
$ cat class.pth
!src/main/php
src/test/php
?vendor/autoload.php
```

Before composer run:

```
$ xp -v
SCAN .
`- PATH class.pth
   `- ADD overlay C:\tools\cygwin\home\Timm\devel\xp\handlebars\src\main\php\
   `- ADD local C:\tools\cygwin\home\Timm\devel\xp\handlebars\src\test\php\

SCAN C:\tools\cygwin\home\Timm\devel\xp\core
`- CORE C:\tools\cygwin\home\Timm\devel\xp\core\src\main\php\lang.base.php
`- PATH boot.pth
   `- ADD core C:\tools\cygwin\home\Timm\devel\xp\core\src\main\php\
   `- ADD core C:\tools\cygwin\home\Timm\devel\xp\core\src\test\php\
   `- ADD core C:\tools\cygwin\home\Timm\devel\xp\core\src\main\resources\
   `- ADD core C:\tools\cygwin\home\Timm\devel\xp\core\src\test\resources\

LOCAL .

array(5) {
  ["base"]=>
  string(66) "C:\tools\cygwin\home\Timm\devel\xp\core\src\main\php\lang.base.php"
  ["overlay"]=>
  array(1) {
    [0]=>
    string(59) "C:\tools\cygwin\home\Timm\devel\xp\handlebars\src\main\php\"
  }
  ["core"]=>
  array(4) {
    [0]=>
    string(53) "C:\tools\cygwin\home\Timm\devel\xp\core\src\main\php\"
    [1]=>
    string(53) "C:\tools\cygwin\home\Timm\devel\xp\core\src\test\php\"
    [2]=>
    string(59) "C:\tools\cygwin\home\Timm\devel\xp\core\src\main\resources\"
    [3]=>
    string(59) "C:\tools\cygwin\home\Timm\devel\xp\core\src\test\resources\"
  }
  ["local"]=>
  array(2) {
    [0]=>
    string(59) "C:\tools\cygwin\home\Timm\devel\xp\handlebars\src\test\php\"
    [1]=>
    string(46) "C:\tools\cygwin\home\Timm\devel\xp\handlebars\"
  }
  ["files"]=>
  array(0) {
  }
}
```

**Comment by thekid** [comment]
## Example for running outside of XP runners

```
$ php -ddate.timezone=Europe/Berlin -r 'require("vendor/autoload.php"); foreach (\lang\ClassLoader::getLoaders() as $l) { \util\cmd\Console::writeLine($l); }'
FileSystemCL<./vendor/xp-framework/logging/src/main/php>
FileSystemCL<./vendor/xp-framework/unittest/src/main/php>
FileSystemCL<./vendor/xp-framework/core/src/main/php>
FileSystemCL<./vendor/xp-framework/core/src/main/resources>
FileSystemCL<./vendor/xp-framework/io-collections/src/main/php>
FileSystemCL<./vendor/xp-forge/mustache/src/main/php>
FileSystemCL<./src/main/php>
```

**Comment by thekid** [comment]
## Path file entries' command line equivalent

| Path flle entry | Command line arg | Arg currently works |
| --- | --- | --- |
| src/main/php | `-cp src/main/php` | :white_check_mark: |
| ~ | `-cp ~` / `-cp $(cygpath -w $HOME)` / `-cp %HOME%` | :white_check_mark: |
| ../lib/entry.xar | `-cp ../lib/entry.xar` | :white_check_mark: |
| /library/asm.xar | `-cp /libary/asm.xar` | :white_check_mark: |
| vendor/autoload.php | `-cp vendor/autoload.php` | :x: |
| !{entry} | `-cp! {entry}` | :x: |
| ?{entry} | `-cp? {entry}` | :x: |

**Comment by thekid** [comment]
- [x] Added support for vendor/autoload.php in https://github.com/xp-framework/xp-runners/pull/39

**Comment by thekid** [comment]
- [x] Added support for "-cp?" and "-cp!" in xp-runners/reference@767a07b

**Comment by thekid** [comment]
- [x] Released XP runners [6.3.0](https://github.com/xp-framework/xp-runners/releases/tag/v6.3.0)

**Comment by thekid** [comment]
With this in place, https://github.com/xp-framework/command/pull/2#issuecomment-170300390 now works seamlessly.
